File: src/lib/slugify.ts
/**
 * Slugify a heading-ish string for use as a filesystem name. Strips
 * leading "#"s, trims, caps at `max` chars, replaces filesystem-unsafe
 * characters with `-`. Returns an empty string when the input is blank.
 */
export function slugifyForFilename(input: string, max = 80): string {
  const cleaned = input.replace(/^#+\s+/, "").trim();
  // Truncate by code point, not code unit — a plain `.slice(0, max)` can sever
  // a surrogate pair and emit a lone surrogate (an invalid character that
  // corrupts the resulting filename) when an emoji/astral char straddles the
  // cap.
  return [...cleaned]
    .slice(0, max)
    .join("")
    .replace(/[/\\:*?"<>|]/g, "-");
}

/** Extract the text of the first H1 (`# Foo`) from a markdown string,
 * or null when none is present in the first 200 lines. */
export function firstHeadingText(md: string): string | null {
  const lines = md.split("\n").slice(0, 200);
  for (const l of lines) {
    if (/^#\s+\S/.test(l)) return l.replace(/^#\s+/, "");
  }
  return null;
}

File: ios/MarkupKit/Sources/MarkupKit/Slugify.swift
import Foundation

/// Slugify a heading-ish string for use as a filesystem name. Strips leading
/// `#`s, trims, caps at `max` chars, replaces filesystem-unsafe characters
/// with `-`. Returns an empty string when the input is blank.
///
/// Faithful port of `src/lib/slugify.ts`.
public func slugifyForFilename(_ input: String, max: Int = 80) -> String {
    var s = input.replacingOccurrences(
        of: "^#+\\s+", with: "", options: .regularExpression)
    s = s.trimmingCharacters(in: .whitespacesAndNewlines)
    if s.count > max { s = String(s.prefix(max)) }
    let unsafe: Set<Character> = ["/", "\\", ":", "*", "?", "\"", "<", ">", "|"]
    return String(s.map { unsafe.contains($0) ? "-" : $0 })
}

/// Extract the text of the first H1 (`# Foo`) from a markdown string, or `nil`
/// when none is present in the first 200 lines. Only a single `#` followed by
/// whitespace counts as an H1.
public func firstHeadingText(_ md: String) -> String? {
    for line in md.components(separatedBy: "\n").prefix(200) {
        if line.range(of: "^#\\s+\\S", options: .regularExpression) != nil {
            return line.replacingOccurrences(
                of: "^#\\s+", with: "", options: .regularExpression)
        }
    }
    return nil
}
